agent/php_user_instrument_wraprec_hashmap.c

```diff
@@ -0,0 +1,485 @@
+    return NULL;
+  }
+
+  hash = nr_scope_hashmap_hash_key(hashmap->log2_num_buckets, key);
+  if (nr_scope_hashmap_fetch_internal(hashmap, hash, key, &bucket)) {
+    return bucket->scoped_funcs_ht;
+  }
+
+  return nr_scope_hashmap_add_internal(hashmap, hash, key);
+}
+
+static void nr_scope_hashmap_destroy_bucket_internal(nr_scope_bucket_t** bucket_ptr) {
+  nr_scope_bucket_t* bucket = *bucket_ptr;
+  nr_free(bucket->key->name);
+  nr_free(bucket->key);
+  nr_func_hashmap_destroy_internal(&bucket->scoped_funcs_ht);
+  nr_realfree((void**)bucket_ptr);
+}
+
+static void nr_scope_hashmap_destroy_internal(nr_scope_hashmap_t** hashmap_ptr) {
+  size_t count;
+  nr_scope_hashmap_t* hashmap;
+  size_t i;
+
+  if ((NULL == hashmap_ptr) || (NULL == *hashmap_ptr)) {
+    return;
+  }
+  hashmap = *hashmap_ptr;
+
+  count = (size_t)(1 << hashmap->log2_num_buckets);
+  for (i = 0; i < count; i++) {
+    nr_scope_bucket_t* bucket = hashmap->buckets[i];
+
+    while (bucket) {
+      nr_scope_bucket_t* next = bucket->next;
+
+      nr_scope_hashmap_destroy_bucket_internal(&bucket);
+
+      bucket = next;
+    }
+  }
+
+  nr_free(hashmap->buckets);
+  nr_realfree((void**)hashmap_ptr);
+}
+
+nr_func_hashmap_t* global_funcs_ht = NULL;
+nr_scope_hashmap_t* scope_ht = NULL;
+
+static void nr_php_user_instrument_wraprec_hashmap_name2keys(
+  nr_func_hashmap_key_t* func,
+  nr_scope_hashmap_key_t* scope,
+  const char* full_name,
+  int full_name_len) {
+
+  if (nrunlikely(NULL == full_name)) {
+    return;
+  }
+  if (nrunlikely(full_name_len <= 0)) {
+    return;
+  }
+
+  func->name = full_name;
+  func->name_len = full_name_len;
+  func->name_hash = 0;
+  scope->name = NULL;
+  scope->name_len = 0;
+  scope->name_hash = 0;
+
+  /* If scope::method, then break into two strings */
+  for (int i = 0; i < full_name_len; i++) {
+    if ((':' == full_name[i]) && (':' == full_name[i + 1])) {
+      func->is_method = true;
+      scope->name = full_name;
+      scope->name_len = i;
+      func->name = full_name + i + 2;
+      func->name_len = full_name_len - i - 2;
+    }
+  }
+
+  if (func->is_method) {
+    scope->name_hash = zend_hash_func(scope->name, scope->name_len);
+  }
+  func->name_hash = zend_hash_func(func->name, func->name_len);
+}
+
+nr_status_t nr_php_user_instrument_wraprec_hashmap_init(void) {
+  if (NULL == scope_ht) {
+    scope_ht = nr_scope_hashmap_create_internal(0);
+  }
+  if (NULL == global_funcs_ht) {
+    global_funcs_ht = nr_func_hashmap_create_internal(0);
+  }
+  if (NULL == scope_ht || NULL == global_funcs_ht) {
+    return NR_FAILURE;
+  }
+  return NR_SUCCESS;
+}
+
+nruserfn_t* nr_php_user_instrument_wraprec_hashmap_add(const char* namestr, size_t namestrlen) {
+  nr_scope_hashmap_key_t scope_key = {0};
+  nr_func_hashmap_key_t func_key = {0};
+  nr_func_hashmap_t* funcs_ht = NULL;
+  bool is_new_wraprec = false;
+  nruserfn_t* wraprec = NULL;
+
+
+  if (NULL == scope_ht || NULL == global_funcs_ht) {
+    return NULL;
+  }
+
```

Comment:
Since it's no longer validating `namestr` here, can you please leave a comment that this function expects `namestr` to be a prevalidated non-null, non-zero length string that doesn't end in `:`?

Reply:
See 575aa95bd52a225774fb2428fa7520d64fdea37f.